### PR TITLE
Move from Sphinx Autodoc to sphinx-autoapi

### PR DESCRIPTION
**Pull request by ptrendx**
As we move forward with the support for multiple frameworks, the behavior of autodoc to import everything becomes an issue - we would need to have a container with all frameworks installed, which is going to be an issue with handling all of the dependencies. AutoAPI does not need to import the module to generate documentation, which is much better in this case.

Also add GitHub action to build the documentation.

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
LGTM. The generated docs look sensible.

File: transformer_engine/pytorch/transformer.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 from flash_attn.flash_attn_interface import flash_attn_unpadded_func
 
-from transformer_engine.pytorch import LayerNormLinear, Linear, LayerNormMLP, LayerNorm
+from transformer_engine.pytorch.module import LayerNormLinear, Linear, LayerNormMLP, LayerNorm
```
Comment (ksivaman) [inline transformer_engine/pytorch/transformer.py:16]: why is this needed?

Comment (ptrendx) [reply transformer_engine/pytorch/transformer.py:16]: Without it autoapi complained about the circular reference since te.pytorch imports stuff from the.pytorch.transformer which then imports things from te.pytorch. I believe we were just lucky before with the order with which we do stuff in __init__.py but we should not rely on it.

**Review by ksivaman — APPROVED** [review]
LGTM
